File: qt/multimedia/gen_qaudiobufferoutput.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QAudioBufferOutput struct {
	h          uintptr
	isSubclass bool
}

// NewQAudioBufferOutput constructs a new QAudioBufferOutput object.
func NewQAudioBufferOutput() *QAudioBufferOutput {
	ret := newQAudioBufferOutput(QAudioBufferOutput_new())
	ret.isSubclass = true
	return ret
}

// NewQAudioBufferOutput2 constructs a new QAudioBufferOutput object.
func NewQAudioBufferOutput2(format *QAudioFormat) *QAudioBufferOutput {
	ret := newQAudioBufferOutput(QAudioBufferOutput_new2(format.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQAudioBufferOutput3 constructs a new QAudioBufferOutput object.
func NewQAudioBufferOutput3(parent *qt.QObject) *QAudioBufferOutput {
	ret := newQAudioBufferOutput(QAudioBufferOutput_new3((*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

// NewQAudioBufferOutput4 constructs a new QAudioBufferOutput object.
func NewQAudioBufferOutput4(format *QAudioFormat, parent *qt.QObject) *QAudioBufferOutput {
	ret := newQAudioBufferOutput(QAudioBufferOutput_new4(format.cPointer(), (*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

func (this *QAudioBufferOutput) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QAudioBufferOutput_MetaObject(this.h)))
}

func (this *QAudioBufferOutput) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QAudioBufferOutput_Metacast(this.h, param1_Cstring))
}

func QAudioBufferOutput_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QAudioBufferOutput_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QAudioBufferOutput) Format() *QAudioFormat {
	_goptr := newQAudioFormat(QAudioBufferOutput_Format(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAudioBufferOutput) AudioBufferReceived(buffer *QAudioBuffer) {
	QAudioBufferOutput_AudioBufferReceived(this.h, buffer.cPointer())
}

func (this *QAudioBufferOutput) OnAudioBufferReceived(slot func(buffer *QAudioBuffer)) {
	QAudioBufferOutput_connect_AudioBufferReceived(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioBufferOutput_AudioBufferReceived
func miqt_exec_callback_QAudioBufferOutput_AudioBufferReceived(cb intptr_t, buffer *QAudioBuffer) {
	gofunc, ok := cgo.Handle(cb).Value().(func(buffer *QAudioBuffer))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQAudioBuffer(buffer)

	gofunc(slotval1)
}

func QAudioBufferOutput_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QAudioBufferOutput_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QAudioBufferOutput_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QAudioBufferOutput_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QAudioBufferOutput) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QAudioBufferOutput_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QAudioBufferOutput) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioBufferOutput_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioBufferOutput_MetaObject
func miqt_exec_callback_QAudioBufferOutput_MetaObject(self QAudioBufferOutput, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QAudioBufferOutput{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QAudioBufferOutput) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QAudioBufferOutput_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QAudioBufferOutput) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioBufferOutput_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioBufferOutput_Metacast
func miqt_exec_callback_QAudioBufferOutput_Metacast(self QAudioBufferOutput, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QAudioBufferOutput{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/multimedia/gen_qaudiobufferoutput.cpp
// +build ignore

#include <QAudioBuffer>
#include <QAudioBufferOutput>
#include <QAudioFormat>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qaudiobufferoutput.h>
#include "gen_qaudiobufferoutput.h"
class MiqtVirtualQAudioBufferOutput : public virtual QAudioBufferOutput {
public:
MiqtVirtualQAudioBufferOutput(): QAudioBufferOutput() {};
MiqtVirtualQAudioBufferOutput(const QAudioFormat& format): QAudioBufferOutput(format) {};
MiqtVirtualQAudioBufferOutput(QObject* parent): QAudioBufferOutput(parent) {};
MiqtVirtualQAudioBufferOutput(const QAudioFormat& format, QObject* parent): QAudioBufferOutput(format, parent) {};

virtual ~MiqtVirtualQAudioBufferOutput() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QAudioBufferOutput::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QAudioBufferOutput_MetaObject(const_cast<MiqtVirtualQAudioBufferOutput*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QAudioBufferOutput::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QAudioBufferOutput::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QAudioBufferOutput_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QAudioBufferOutput::qt_metacast(param1);

	}
};
QAudioBufferOutput* QAudioBufferOutput_new() {
return new MiqtVirtualQAudioBufferOutput();
}
QAudioBufferOutput* QAudioBufferOutput_new2(QAudioFormat* format) {
return new MiqtVirtualQAudioBufferOutput(*format);
}
QAudioBufferOutput* QAudioBufferOutput_new3(QObject* parent) {
return new MiqtVirtualQAudioBufferOutput(parent);
}
QAudioBufferOutput* QAudioBufferOutput_new4(QAudioFormat* format, QObject* parent) {
return new MiqtVirtualQAudioBufferOutput(*format, parent);
}
void QAudioBufferOutput_virtbase(QAudioBufferOutput* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QAudioBufferOutput_MetaObject(const QAudioBufferOutput* self) {
	return (QMetaObject*) self->metaObject();
}
void* QAudioBufferOutput_Metacast(QAudioBufferOutput* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QAudioBufferOutput_Tr(const char* s) {
	QString _ret = QAudioBufferOutput::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
QAudioFormat* QAudioBufferOutput_Format(const QAudioBufferOutput* self) {
	return new QAudioFormat(self->format());
}
void QAudioBufferOutput_AudioBufferReceived(QAudioBufferOutput* self, QAudioBuffer* buffer) {
	self->audioBufferReceived(*buffer);
}
void QAudioBufferOutput_connect_AudioBufferReceived(QAudioBufferOutput* self, intptr_t slot) {
	MiqtVirtualQAudioBufferOutput::connect(self, static_cast<void (QAudioBufferOutput::*)(const QAudioBuffer&)>(&QAudioBufferOutput::audioBufferReceived), self, [=](const QAudioBuffer& buffer) {
		const QAudioBuffer& buffer_ret = buffer;
		// Cast returned reference into pointer
		QAudioBuffer* sigval1 = const_cast<QAudioBuffer*>(&buffer_ret);
		miqt_exec_callback_QAudioBufferOutput_AudioBufferReceived(slot, sigval1);
	});
}
struct miqt_string QAudioBufferOutput_Tr2(const char* s, const char* c) {
	QString _ret = QAudioBufferOutput::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QAudioBufferOutput_Tr3(const char* s, const char* c, int n) {
	QString _ret = QAudioBufferOutput::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QAudioBufferOutput_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQAudioBufferOutput*>( (QAudioBufferOutput*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QAudioBufferOutput_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQAudioBufferOutput*)(self) )->virtualbase_MetaObject();
}
void QAudioBufferOutput_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQAudioBufferOutput*>( (QAudioBufferOutput*)(self) )->handle__Metacast = slot;
}
void* QAudioBufferOutput_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQAudioBufferOutput*)(self) )->virtualbase_Metacast(param1);
}
void QAudioBufferOutput_Delete(QAudioBufferOutput* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQAudioBufferOutput*>( self );
	} else {
		delete self;
	}
}
